finality-aleph/src/testing/network.rs

```diff
@@ -0,0 +1,315 @@
+            {
+                assert_eq!(protocol, Protocol::Generic.name());
+                sent_auth.insert(peer_id, auth_data);
+            }
+        }
+
+        let mut expected_auth = HashMap::new();
+        for authority in self.authorities.iter().skip(1) {
+            expected_auth.insert(authority.peer_id(), authentication.clone());
+        }
+
+        assert_eq!(sent_auth, expected_auth);
+    }
+
+    async fn connect_session_authorities(&mut self, session_id: u32) {
+        for (index, authority) in self.authorities.clone().into_iter().enumerate().skip(1) {
+            let handler = self.get_session_handler(index, session_id).await;
+
+            self.connect_identity_to_network(authority.peer_id(), Protocol::Generic);
+            self.connect_identity_to_network(authority.peer_id(), Protocol::Validator);
+
+            self.network.emit_event(Event::NotificationsReceived {
+                remote: authority.peer_id().into(),
+                messages: vec![(
+                    Protocol::Generic.name(),
+                    NetworkData::<MockData>::Meta(DiscoveryMessage::AuthenticationBroadcast(
+                        handler.authentication().unwrap(),
+                    ))
+                    .encode()
+                    .into(),
+                )],
+            });
+        }
+    }
+
+    async fn start_session(&mut self, session_id: u32) -> SessionNetwork<MockData> {
+        let data_network = self.start_validator_session(0, session_id).await;
+        self.connect_session_authorities(session_id).await;
+        self.check_sends_add_reserved_node().await;
+        self.check_sends_authentication(
+            self.get_session_handler(0, session_id)
+                .await
+                .authentication()
+                .unwrap(),
+        )
+        .await;
+
+        data_network
+    }
+
+    fn emit_notifications_received(
+        &mut self,
+        node_id: usize,
+        messages: Vec<NetworkData<MockData>>,
+    ) {
+        self.network.emit_event(Event::NotificationsReceived {
+            remote: self.authorities[node_id].peer_id().into(),
+            messages: messages
+                .iter()
+                .map(|m| (Protocol::Validator.name(), m.encode().into()))
+                .collect(),
+        });
+    }
+
+    async fn cleanup(self) {
+        self.network_manager_exit_tx.send(()).unwrap();
+        self.network_service_exit_tx.send(()).unwrap();
+        self.network_manager_handle.await.unwrap();
+        self.network_service_handle.await.unwrap();
+        while self.network.send_message.try_next().await.is_some() {}
+        self.network.close_channels().await;
+    }
+}
+
+#[tokio::test]
+async fn test_connects_to_others() {
+    let mut test_data = prepare_one_session_test_data().await;
+    let mut data_network = test_data.start_session(43).await;
+
+    test_data.emit_notifications_received(1, vec![NetworkData::Data(vec![1, 2, 3], SessionId(43))]);
```

Comment:
declare `43` and `vec![1, 2, 3]` as variables for clarity

Reply:
Done